homeassistant/components/guardian/__init__.py

```diff
@@ -142,20 +284,135 @@ def icon(self) -> str:
+        return False
+
+    @callback
+    def _async_update_from_latest_data(self):
+        """Update the entity.
+
+        This should be extended by Guardian platforms.
+        """
+        raise NotImplementedError
+
+    @callback
+    def _async_update_state_callback(self):
+        """Update the entity's state."""
+        self._async_update_from_latest_data()
+        self.async_write_ha_state()
+
+
+class PairedSensorEntity(GuardianEntity):
+    """Define a Guardian paired sensor entity."""
+
+    def __init__(
+        self,
+        entry: ConfigEntry,
+        coordinator: DataUpdateCoordinator,
+        kind: str,
+        name: str,
+        device_class: str,
+        icon: str,
+    ) -> None:
+        """Initialize."""
+        super().__init__(entry, kind, name, device_class, icon)
+
+        self._coordinator = coordinator
+        self._paired_sensor_uid = coordinator.data["uid"]
+
+        self._device_info["identifiers"] = {(DOMAIN, self._paired_sensor_uid)}
+        self._device_info["name"] = f"Guardian Paired Sensor {self._paired_sensor_uid}"
+        self._device_info["via_device"] = (DOMAIN, self._entry.data[CONF_UID])
+
     @property
     def name(self) -> str:
         """Return the name of the entity."""
-        return f"Guardian {self._valve_controller_uid}: {self._name}"
+        return f"Guardian Paired Sensor {self._paired_sensor_uid}: {self._name}"
 
     @property
-    def should_poll(self) -> bool:
-        """Return True if entity has to be polled for state."""
-        return False
+    def unique_id(self):
+        """Return the unique ID of the entity."""
+        return f"{self._paired_sensor_uid}_{self._kind}"
+
+    async def async_added_to_hass(self) -> None:
+        """Perform tasks when the entity is added."""
+        self.async_on_remove(
+            self._coordinator.async_add_listener(self._async_update_state_callback)
+        )
+
+        async def async_remove(uid: str):
+            """Remove the entity from the registry and the state machine.
+
+            Also removes the associated paired sensor device if this is the last entity.
+            """
+            if uid != self._paired_sensor_uid:
+                return
+
+            ent_reg = await self.hass.helpers.entity_registry.async_get_registry()
+            ent_entry = ent_reg.async_get(self.entity_id)
+            if not ent_entry:
+                await self.async_remove()
+                return
+
+            dev_reg = await self.hass.helpers.device_registry.async_get_registry()
+            dev_ent = dev_reg.async_get(ent_entry.device_id)
+            if not dev_ent:
+                ent_reg.async_remove(self.entity_id)
+                return
+
+            if len(async_entries_for_device(ent_reg, ent_entry.device_id)) == 1:
+                dev_reg.async_remove_device(dev_ent.id)
+                return
+
+            ent_reg.async_remove(self.entity_id)
```

Comment:
This is not necessary. Just remove the device and it will clean up the entity registry AND the entities. And that can just be done where this signal is received, should not happen in the entity